### DISCUSSION TITLE
Built paste command with sudo option

### DISCUSSION THREAD

**Discussion by og900aero** [gokcehan/lf#1227] [Q&A]
The built-in copy command works very well, but I would often copy to a location that requires root access. That's why I made a custom paste command, which I call in the following way when copying:

```
map p :copy; paste_custom

cmd paste_custom &{{
	if sudo -u "$USER" test -w "$PWD"; then
	    lf -remote "send paste"
	else
	    sudo -A lf -remote "send paste"
	fi
	lf -remote "send clear"
	lf -remote "send unselect"
	lf -remote "send reload"
}}
```

The problem is that it asks for the password, but the copy itself does not happen. sudo -A definitely works, because if I enter it in a terminal with a cp command, it runs fine. What could be the problem?

**Comment by joelim-work** [comment]
**TL;DR:** You need to start the main instance of `lf` with root permissions too. **But it is recommended to instead use `mv`/`cp` to perform the actual file move/copy operation.**

---

There's three instances of `lf` involved in your example:

- `lf` - the main instance that you work with
- `lf -server` a server running in the background that manages the main instances
- `lf -remote` used for sending commands

The way `lf -remote` works is that it simply sends a command string like `send paste` to the server and then terminates. The server then forwards the command to the main instance, which will do the actual work.

The instances of `lf` communicate by using a socket, usually located at a place like `/tmp/lf.<user>.sock`. Note that there is a different socket for each user. This means that if you run `lf -remote` with root permissions, then it would only be able to communicate if the main instance and the server are also running with root permissions. Also, the main instance wouldn't be able to physically copy the file anyway if it wasn't running with root permissions.

---

**Disclaimer:** The usual warnings for running programs under `sudo` applies.

**Reply by og900aero** [reply]
> **TL;DR:** You need to start the main instance of `lf` with root permissions too.

Wrong, no need rooted lf instance. I write the following code under lf without problem:

```
cmd paste_custom &{{
	if sudo -u "$USER" test -w "$PWD"; then
	    lf -remote "send paste"
	else
	    set -- $(cat ~/.local/share/lf/files)
	    mode="$1"
	    shift
	    case "$mode" in
	        copy) sudo -A cp -rn -- "$@" .;;
	        move) sudo -A mv -n -- "$@" .;;
	    esac
	    rm ~/.local/share/lf/files
	fi
	lf -remote "send clear"
	lf -remote "send unselect"
	lf -remote "send reload"
}}
```
The problem with this solution is that it does not use the built-in progress function with copy.

**Reply by joelim-work** [reply]
Ah OK, I was under the impression you wanted to use `lf` itself to perform the actual move/copy operation. I do agree that using `mv`/`cp` is a much safer approach, just as you have done. 👍 

If it's not too much trouble, perhaps you can add your example to the [Tips](https://github.com/gokcehan/lf/wiki/Tips) section in the wiki.  I don't see any mention of anything similar, and I'm sure your snippet will help other users looking for the same thing. Although there's already quite a few variations on moving/copying files, and the only difference here is the use of `sudo`.
